### PR TITLE
Add dest parameter to filter option in copy and copySync

### LINKED ISSUE #351: Additional param for options.filter() on copy()
At the moment when using `copy()`, the `options.filter()` function receives the source `path` as it's only param. It would be nice to have both the source and destination paths:

```javascript
fs.copy(srcFolder, destFolder, {
  filter: (srcPath, destPath) => {
    return true
  }
}, cb)
```

This would make checking last mod dates much easier.

### PR DESCRIPTION
**Pull request by manidlou**
This resolves #351.

I added dest as an additional parameter to filter option in both `copy` and `copySync`.

### CODE REVIEW

File: lib/copy-sync/__tests__/copy-sync-dir.test.js [hunk 1]
```diff
@@ -121,29 +147,61 @@ describe('+ copySync()', () => {
 
       assert(!fs.existsSync(path.join(dest, IGNORE)), 'directory was not ignored')
       assert(!fs.existsSync(path.join(dest, IGNORE, 'file')), 'file was not ignored')
+      done()
```
Comment (RyanZim) [inline lib/copy-sync/__tests__/copy-sync-dir.test.js]: Revert this change. Using `done` isn't needed since we are using sync methods here.

File: lib/copy-sync/__tests__/copy-sync-dir.test.js [hunk 2]
```diff
@@ -104,9 +129,10 @@ describe('+ copySync()', () => {
           assert(!fs.existsSync(path.join(destSub, j.toString())))
         }
       }
+      done()
```
Comment (RyanZim) [inline lib/copy-sync/__tests__/copy-sync-dir.test.js]: Revert this change. Using `done` isn't needed since we are using sync methods here.

File: lib/copy-sync/copy-sync.js [hunk 3]
```diff
@@ -35,8 +35,8 @@ function copySync (src, dest, options) {
 
   if (options.filter instanceof RegExp) {
     console.warn('Warning: fs-extra: Passing a RegExp filter is deprecated, use a function')
-    performCopy = options.filter.test(src)
-  } else if (typeof options.filter === 'function') performCopy = options.filter(src)
+    performCopy = options.filter.test(src, dest)
```
Comment (RyanZim) [inline lib/copy-sync/copy-sync.js]: This change doesn't do anything; `Regex.prototype.test` only accepts one param.

File: lib/copy/ncp.js [hunk 4]
```diff
@@ -37,11 +37,11 @@ function ncp (source, dest, options, callback) {
     if (filter) {
       if (filter instanceof RegExp) {
         console.warn('Warning: fs-extra: Passing a RegExp filter is deprecated, use a function')
-        if (!filter.test(source)) {
+        if (!filter.test(source, dest)) {
```
Comment (RyanZim) [inline lib/copy/ncp.js]: This change doesn't do anything; `Regex.prototype.test` only accepts one param.

File: lib/copy-sync/copy-sync.js (hunk 3 above)
Comment (manidlou) [inline lib/copy-sync/copy-sync.js]: Very true. I missed it. Thanks @RyanZim. Great that you caught this.

**Comment by manidlou** [comment]
I fixed them. Please let me know if there is any other issues left.

File: lib/copy-sync/__tests__/copy-sync-dir.test.js [hunk 5]
```diff
@@ -147,50 +146,47 @@ describe('+ copySync()', () => {
...
+        fs.mkdirsSync(src)
+        fs.writeFileSync(path.join(src, 'somefile.html'), 'some data')
         fs.mkdirsSync(dest)
         try {
           fs.copySync(src, dest, filter)
         } catch (err) {
           assert.ifError(err)
         }
-        assert(!fs.existsSync(path.join(dest, 'subdir')))
-        done()
```
Comment (RyanZim) [inline lib/copy-sync/__tests__/copy-sync-dir.test.js]: You still need `done` when you're using setTimeout (since that makes the code async).

**Comment by manidlou** [comment]
Should I merge this?

**Comment by RyanZim** [comment]
@manidlou It's up to you. I was leaving it for @jprichardson, since I'm a bit tired today and could have missed something.

---

@jprichardson Perhaps we should formalize some collaborator guidelines sometime?

File: lib/copy-sync/__tests__/copy-sync-dir.test.js [hunk 6]
```diff
@@ -27,15 +27,15 @@ describe('+ copySync()', () => {
       src = path.join(TEST_DIR, 'src')
       dest = path.join(TEST_DIR, 'dest')
 
-      fs.mkdirsSync(src)
+      fs.mkdirSync(src)
```
Comment (jprichardson) [inline lib/copy-sync/__tests__/copy-sync-dir.test.js]: Why the conversion here to `mkdirSync` from `mkdirsSync`?

Comment (RyanZim) [inline lib/copy-sync/__tests__/copy-sync-dir.test.js]: I can't speak for @manidlou, but `mkdirs` technically isn't needed, since we know the parent directory exists.

Comment (manidlou) [inline lib/copy-sync/__tests__/copy-sync-dir.test.js]: > @jprichardson Perhaps we should formalize some collaborator guidelines sometime? :+1: 

> Why the conversion here to mkdirSync from mkdirsSync?

As @RyanZim mentioned, I thought we wouldn't need to call `mkdirs` since we know the parent directory exist. So, I guess I should revert to `mkdirsSync`. Is that right?

And also sorry that I am still not in harmony with you guys. I try my best to be as helpful as possible.

Comment (jprichardson) [inline lib/copy-sync/__tests__/copy-sync-dir.test.js]: > And also sorry that I am still not in harmony with you guys. I try my best to be as helpful as possible.

@manidlou you're very helpful, thank you!!

> @jprichardson Perhaps we should formalize some collaborator guidelines sometime?

I think for now, any modifications (including my own), barring some crazy necessary hot-fix, require at least one other reviewer? 

To be clear, it's okay to accept a self-submitted PR, but it must be for a very good reason. We're all professionals here, so I think this loose guideline should be acceptable for now.

Sound good @manidlou @RyanZim?

**Comment by jprichardson** [comment]
Let me know if there is anything else that's needed before I publish 2.1.0.

**Comment by manidlou** [comment]
Sounds good.

Just a quick question, for issues like #358 that user doesn't experience the issue anymore, is this a correct action to comment something like "closing the issue" and then closing it?

By the way, I close this branch as it is already merged.

**Comment by RyanZim** [comment]
> And also sorry that I am still not in harmony with you guys. I try my best to be as helpful as possible.

You're doing great.

> Just a quick question, for issues like #358 that user doesn't experience the issue anymore, is this a correct action to comment something like "closing the issue" and then closing it?

Generally, if the user can't reproduce, then yes, you would comment & close. #358 is a gray area, since it is a problem on other Node versions. I'm personally not sure what to do here. @jprichardson?
